---
title: "class13"
author: "Negin Silani"
date: "5/14/2019"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#Sample genotypes in the MXL 1000 Genome project data

Here we focus on the Mexican Ancestory in Los Anglees CA (MXL) Popilation 

What proportion of the Mexican Ancestry in Los Angeles sample population (MXL)
are homozygous for the asthma associated SNP (G|G) rs8067378

```{r}
#Read CSV from ENSEMBLE
mxl<- read.csv("373531-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378 (1).csv")
head(mxl)
```

How many of the genotypes are there?
```{r}
table(mxl$Genotype..forward.strand.)
```

Proportion or percent of total each genotype

```{r}
(table(mxl$Genotype..forward.strand.)/ nrow(mxl) )*100
```

## Quality Scores in FASTQ files

The forth line of a FASTQ sequence format file encodes the quality score that tells us how good the sequence aat a given position 

```{r}
library(seqinr)
library(gtools)

phred <- asc( s2c("DDDDCDEDCDDDDBBDDDCC@") ) - 33
#phred
```


```{r}
read.table("rs8067378_ENSG00000172057.6.txt")
```

```{r}
summary("rs8067378_ENSG00000172057.6.txt")
```

##Population Scale analysis

Read RNA-seq count data with genotype information results table

```{r}
expr<- read.table("rs8067378_ENSG00000172057.6.txt")

head(expr)
```

```{r}
summary(expr)
```

```{r}
inds <- expr$geno == "G/G"
summary(expr[inds,"exp"])
```

```{r}
inds <- expr$geno == "A/G"
summary(expr[inds,"exp"])
```

```{r}
inds <- expr$geno == "A/A"
summary(expr[inds,"exp"])
```


```{r}
boxplot(exp ~ geno, data=expr)
```


